### PR TITLE
Fixes #19393 - move upload report to yum plugin

### PR DESCRIPTION
**Pull request by jlsherrill**
In addition this adds --force options to both the enabled
repo report and the package profile upload

### CODE REVIEW

**Comment by jlsherrill** [comment]
scratch build:  http://koji.katello.org/koji/taskinfo?taskID=634096

**Comment by jlsherrill** [comment]
Notes about change and what needs to be tested:

1) katello-package-upload --force  now should upload the package profile even if no change has occurred (you should be able to monitor the server logs to verify this)
2) errata applicability now works even when goferd has never run
3) errata applicability is updated if you disable a repository in the ui for a host, run 'subscription-manager refresh && yum repolist'
4) errata applicability is updated if you disable or enable a repository using 'subscription-manager repos'
5) errata applicability is updated if you using 'yum-config-manager to enable or disable a repo
6) the enabled repos report is not actually sent to the server on every yum action, but only on a change.  
7) katello-enabled-repos-upload can be used to manually upload the report
8) katello-enabled-repos upload --force can be used to force the upload of the report even if it doesn't think it needs to be
9) if the server is not registered, yum commands should not error
10) if the server is registered, but has had its profile deleted, yum commands should not error

**Comment by bbuckingham** [comment]
From testing, a couple of observations:

4. errata applicability is updated if you disable or enable a repository using 'subscription-manager repos'
    NOTE: on my configuration, this required a yum operation to occur (e.g. 'yum repolist')

5. errata applicability is updated if you using 'yum-config-manager to enable or disable a repo
    NOTE: when I ran 'yum-config-manager --disable rhel-7-server-rpms', it disabled the repo locally (redhat.repo); however, there was no communication with the server.  Also, performing 'yum repolist', the change did not propagate to the server.  Should it?  (e.g. on Content Hosts, the client still showed the repo as enabled)

9. if the server is not registered, yum commands should not error
    NOTE: The commands were executed, but the user will see the following message:
    'Cannot upload enabled repos report, is this client registered?'

**Comment by bbuckingham** [comment]
One additional difference in behavior that I noticed is that after an initial 'subscription-manager register', the generate errata applicability task is not initiated within katello.  With the gofer plugin, it does appear to get initiated.

**Comment by bbuckingham** [comment]
The code looks good; however, it may be nice to have a second glance by a python developer.  @jortel or @daviddavis, would one of you guys be willing to give it a quick lookover?  Thanks!

File: test/test_yum_plugins/test_enabled_repos_upload.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        fake_certificate.getConsumerId.assert_called_with()
+        fake_report_enabled.assert_called_with(consumer_id, FAKE_REPORT)
+
+    @patch('enabled_repos_upload.EnabledReport')
+    @patch('enabled_repos_upload.ConsumerIdentity.read')
+    @patch('enabled_repos_upload.UEP.report_enabled')
+    @patch('enabled_repos_upload.EnabledRepoCache.is_valid')
+    @patch('enabled_repos_upload.EnabledRepoCache.save')
+    def test_cached(self, cache_save, cache_valid, fake_report_enabled, fake_read, fake_report):
+        #import pdb; pdb.set_trace()
```
Comment (daviddavis) [inline test/test_yum_plugins/test_enabled_repos_upload.py]: Might just want to drop this.

**Comment by daviddavis** [comment]
LGTM

**Comment by jlsherrill** [comment]
Updated to add the enabled repo uploading to goferd to attempt to provide a seamless experience for those users.  new test rpms are here:  http://koji.katello.org/koji/taskinfo?taskID=639783

**Comment by bbuckingham** [comment]
Tested both with and without gofer using the updated rpm.  Behavior looks good; however, I do still see the behaviors mentioned in https://github.com/Katello/katello-agent/pull/44#issuecomment-302511629 on the goferless configuration.  Are these expected?

**Comment by jlsherrill** [comment]
It is expected, whether its acceptable or not, is another question :)

**Comment by bbuckingham** [comment]
Regarding https://github.com/Katello/katello-agent/pull/44#issuecomment-302511629, 

- errata applicability is updated if you disable or enable a repository using 'subscription-manager repos' :
The behavior on this one is different between gofer and yum plugins.  In order to make it consistent, would it make sense to have this be handled by a subscription-manager plugin? 

- errata applicability is updated if you using 'yum-config-manager to enable or disable a repo :
The behavior mentioned on this one appears to be the same regardless of using the gofer or yum plugins; therefore, while it may not be ideal, they are consistent.

**Comment by jlsherrill** [comment]
Regarding https://github.com/Katello/katello-agent/pull/44#issuecomment-302511629, 

> errata applicability is updated if you disable or enable a repository using 'subscription-manager repos' :
The behavior on this one is different between gofer and yum plugins.  In order to make it consistent, would it make sense to have this be handled by a subscription-manager plugin? 

I don't think its possible to create a sub-man plugin to handle this.  The only way i can see to handle this would be to move all this functionality to some other service that is always running and monitoring for file changes like goferd, which i was trying to avoid.

> errata applicability is updated if you using 'yum-config-manager to enable or disable a repo :
The behavior mentioned on this one appears to be the same regardless of using the gofer or yum plugins; therefore, while it may not be ideal, they are consistent.

That seems strange, i'd expect the goferd case to work, is the redhat.repo file being updated?

**Comment by bbuckingham** [comment]
Regarding https://github.com/Katello/katello-agent/pull/44#issuecomment-304030979:

For the first case, perhaps we can move forward without a plugin or service and obtain some user feedback.  Folks that are used to the behavior with gofer may initially stumble until they realize that they need to perform a yum action after enabling/disabling a repo using sub-mgr.

For the second case, it looks like this may be a separate bug within katello itself.  In the gofer case, the enabled_repos API is invoked with an empty list; however, the disabling is not reflected on the katello server.  This may only be an issue when there was only a single repository enabled on the client (which was my case).  Below is an example of the API call:

```
Started PUT "/rhsm/systems/b1f51b23-289b-4b01-b2fb-68fd6aab549f/enabled_repos" for 127.0.0.1 at 2017-05-26 13:22:54 +0000
Processing by Katello::Api::Rhsm::CandlepinProxiesController#enabled_repos as JSON
   Parameters: {"enabled_repos"=>{"repos"=>[]}, "id"=>"b1f51b23-289b-4b01-b2fb-68fd6aab549f"}
```

**Comment by bbuckingham** [comment]
Discussed https://github.com/Katello/katello-agent/pull/44#issuecomment-304284755 with @jlsherrill on IRC.  Although the client will send the 'enabled_repos' API to the server during a 'yum-config-manager --enable' (or disable), it is for the purpose of regenerating the applicability.  In other words, it should not update (on the server) the list of repositories enabled (e.g. Hosts -> Content Hosts -> [myhost] -> Repository Sets).  

Based on that, ACK.  Let's get this thing in to the community for some user feedback. :)

Nice work @jlsherrill !

**Pull request merged**